src/arithmetic/hashtocurve.rs

```diff
@@ -0,0 +1,330 @@
+    ) -> Box<dyn Fn(&[u8]) -> C::Projective + '_> {
+        let domain_separation_tag = format!(
+            "{}-{}_{}_{}_NU_",
+            domain_prefix,
+            C::CURVE_ID,
+            hasher.hash_name(),
+            self.map_id()
+        );
+
+        Box::new(move |message| {
+            let us = hasher.hash_to_field(message, domain_separation_tag.as_bytes(), 1);
+            let r = self.map_to_curve(&us[0]);
+            self.iso_map(&r)
+        })
+    }
+}
+
+/// Method of hashing a message and domain_separation_tag to field elements.
+pub trait MessageHasher<F: FieldExt> {
+    /// The HASH_NAME of this message hasher as specified in
+    /// <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-suite-id-naming-conventions-2>.
+    fn hash_name(&self) -> &str;
+
+    /// Hash the given message and domain separation tag to give `count`
+    /// field elements.
+    fn hash_to_field(&self, message: &[u8], domain_separation_tag: &[u8], count: usize) -> Vec<F>;
+}
+
+/// A MessageHasher for SHAKE128
+/// [FIPS202](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf).
+/// It does not support domain separation tags longer than 128 bytes.
+#[derive(Debug, Default)]
+pub struct Shake128<F: FieldExt> {
+    marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> MessageHasher<F> for Shake128<F> {
+    fn hash_name(&self) -> &str {
+        "XOF:SHAKE128"
+    }
+
+    fn hash_to_field(&self, message: &[u8], domain_separation_tag: &[u8], count: usize) -> Vec<F> {
+        use sha3::digest::{ExtendableOutput, Update};
+        assert!(domain_separation_tag.len() < 256);
+
+        // Assume that the field size is 32 bytes and k is 256, where k is defined in
+        // <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-security-considerations-3>.
+        const CHUNKLEN: usize = 64;
+
+        let outlen = count * CHUNKLEN;
+        let mut outlen_enc = vec![];
+        outlen_enc.write_u32::<BigEndian>(outlen as u32).unwrap();
+
+        let mut xof = sha3::Shake128::default();
+        xof.update(message);
+        xof.update(outlen_enc);
+        xof.update([domain_separation_tag.len() as u8]);
+        xof.update(domain_separation_tag);
+
+        xof.finalize_boxed(outlen)
+            .chunks(CHUNKLEN)
+            .map(|big| {
+                let mut little = [0u8; CHUNKLEN];
+                little.copy_from_slice(big);
+                little.reverse();
+                F::from_bytes_wide(&little)
+            })
+            .collect()
+    }
+}
+
+/// A MessageHasher for BLAKE2b.
+#[derive(Debug, Default)]
+pub struct Blake2bXof<F: FieldExt> {
+    marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> MessageHasher<F> for Blake2bXof<F> {
+    fn hash_name(&self) -> &str {
+        "XOF:BLAKE2b"
+    }
+
+    #[allow(unused_variables)]
+    fn hash_to_field(&self, message: &[u8], domain_separation_tag: &[u8], count: usize) -> Vec<F> {
+        todo!()
```

Comment:
The entire `Blake2bXof` should be removed from this PR, or this should be finished and the SHAKE128 thing should be removed -- we don't need multiple instances of hashers in the codebase right now.

Reply:
I included Shake128 because I was originally thinking that would make it easier to compare to [the secp256k1 test vectors from the Internet Draft](https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#section-appendix.j), but I hadn't noticed that those only have XMD:SHA-256. I do want to have more confidence that we've implemented the same thing as the draft; I'll think about how to do that.

Reply:
I'm going to switch the hasher to XMD:BLAKE2b.